### DISCUSSION TITLE
iPXE + Wireguard == UwU

### DISCUSSION THREAD

**Discussion by suqdiq** [ipxe/ipxe#460] [Ideas]
yes please?

@mcb30 how much queen tokens do you think it would take for you to do it?

KK 10X BAI <3

**Comment by NiKiZe** [comment]
It would be good to include a technical spec, or at least links to some resources. Assume that we know nothing about what Wireguard or UwU is. (please update orignal post with this information so that it is easier to give you good answers)
Also some example on how this feature would be used, as in how would this be configured in iPXE, which commands would be needed to make the configuration, and is there any other commands that would be needed, or just nice to have?

**Comment by assimilat** [comment]
The OP's poor choice of wording aside (UwU is an ascii-ized emoticon), I think there could be a legit use for Wireguard in iPXE. [Wireguard](https://www.wireguard.com/) is an extremely lightweight, low resource, secure VPN system that was originally designed for (and has since been merged into) the linux kernel. It has userspace ports in [Go](https://git.zx2c4.com/wireguard-go/about), [Rust](https://git.zx2c4.com/wireguard-rs/about), an incomplete [Haskell](https://git.zx2c4.com/wireguard-hs/about) port, and there are stable ports to Windows, MacOs, OpenBSD, FreeBSD, iOS, Android(while Android is still technically linux it has some user space components built in as well). A good overview of the OG code can be found in their [backports repository](https://git.zx2c4.com/wireguard-linux-compat/tree/).

Like any VPN system Wireguard needs a private key for the local node and a set of public keys and ip addresses for remote nodes. That would be the tricky bit in terms of orchestration. My thought on how that could be achieved would be to either roll it into an image at compile time or have iPXE look for a pre-existing config at a set path on the host system. The system would have to be preprovisioned to some degree for that to be at all viable of course (or there would need to be another key distribution component in place (maybe at the hypervisor level)).

As for a potential use case, Say for instance you have a VPS/Cloud server provider that doesn't have an image for a linux distribution or operating system that you would like to run remotely, you could build an instance by configuring the hardware resources, install whatever arbitrary OS they provide, place the Wireguard config somewhere on the ESP and chainload iPXE from the existing bootloader.
The order of operations from there would require iPXE to detect and load the local Wireguard configuration, which would give the remote system a secure VPN connection back to a peer that could provide bootable images or scripts to further the systems reprovisioning. And of course if the system re-uses the existing ESP then the bundled wireguard configuration could be used by the replacement host OS as well.
I can probably put together several other use cases if you would like, but the request is essentially just giving iPXE lightweight and fast VPN capabilities.

As for how it could be used, the first thing that comes to mind would be changing the operating systems on VPS servers that are hosted by providers with limited selections.

**Reply by NiKiZe** [reply]
That is source code, not spec, iPXE can handle private certs, still no example on client side setup, this needs to be a number of commands that can run in iPXE or settings/variables

**Reply by assimilat** [reply]
Yep! I was just trying to outline the base concept a bit more than the OP did. I can look into what would be needed later and update.
EDIT:
Here is a whitepaper https://www.wireguard.com/papers/wireguard.pdf
and the quick and dirty explanation: https://www.wireguard.com/protocol/
still going to add more detail when i get a chance.

**Comment by suqdiq** [comment]
so i've finally had some time to work on this!

wireguard in iPXE from the User perspective

primary goal:

- Securely boot systems over untrusted networks (internet, public WiFi) by tunneling boot traffic through WireGuard to a trusted boot server.

Proposed iPXE WireGuard Commands

| Command | Syntax   | Description     |Example        |
|---------|------------------------------------------------------------------------------------------------|------------------------------------|-----------------------------------------------------|
| wgup    | wgup \<interface\>             | Bring up WireGuard interface and establish tunnel        | wgup wg0   |
| wgdown  | wgdown \<interface\>       | Tear down WireGuard interface and clean up resources   | wgdown wg0   |
| wgstat  | wgstat [interface]       | Display WireGuard interface status and statistics  | wgstat wg0  |
| wgconf  | wgconf --interface \<name\> --private-key \<key\> --address \<ip\/cidr\> --peer \<pubkey\> --endpoint \<host:port\> --allowed-ips \<cidrs\> | Configure WireGuard interface in a single command     | wgconf --interface wg0 --private-key ${wg_key} --address 10.0.0.2/24 --peer BASE64KEY  --endpoint vpn.example.com:51820 --allowed-ips 10.0.0.0/24 |
| wgparse | wgparse \<config-image\>      | Parse WireGuard configuration from downloaded image/file | wgparse wgconfig|

  Proposed iPXE WireGuard Settings

  | Setting Path             | Type         | Description                                    | Example Value                                |
  |--------------------------|--------------|------------------------------------------------|----------------------------------------------|
  | wg0/private-key          | string       | Interface private key (base64)                 | oK7Qg3fXgVTVqBJBM5pVzN8gZ7F3qXtLvYqfEp+3n2c= |
  | wg0/public-key           | string (RO)  | Interface public key (auto-derived, read-only) | fE/wdxzl0klVp/IR8UcaoGUMjqaWi3jAd7KzHKFS6Ds= |
  | wg0/address              | string       | Interface IP address with CIDR notation        | 10.0.0.2/24                                  |
  | wg0/port                 | integer      | UDP listen port                                | 51820                                        |
  | wg0/mtu                  | integer      | Interface MTU                                  | 1420                                         |
  | wg0/peer/0/public-key    | string       | Peer's public key (base64)                     | HIgo9xNzJMWLKASShiTqIybxZ0U3wGLiUeJ1PKf8ykw= |
  | wg0/peer/0/endpoint      | string       | Peer's endpoint address                        | vpn.example.com:51820                        |
  | wg0/peer/0/allowed-ips   | string       | Comma-separated list of allowed IP ranges      | 10.0.0.0/24,172.16.0.0/12                    |
  | wg0/peer/0/keepalive     | integer      | Persistent keepalive interval in seconds       | 25                                           |
  | wg0/peer/0/preshared-key | string       | Optional pre-shared key (base64)               | FpCyhws9cxwWoV4xELtfJvjJN+zQVRPISllRWgeopVE= |
  | wg0/status               | string (RO)  | Interface status                               | up, down, handshaking                        |
  | wg0/last-handshake       | integer (RO) | Seconds since last successful handshake        | 30                                           |



Examples

  1. Basic wireguard Configuration via iPXE settings
```
  # Configure wireguard interface settings
  set wg0/private-key BASE64-ENCODED-KEY
  set wg0/address 10.0.0.2/24
  set wg0/port 51820

  # Configure peer (boot server)
  set wg0/peer/0/public-key BASE64-ENCODED-KEY
  set wg0/peer/0/endpoint bootserver.example.com:51820
  set wg0/peer/0/allowed-ips 10.0.0.0/24,192.168.1.0/24
  set wg0/peer/0/keepalive 25

  # Bring up wireguard interface
  wgup wg0
```
  2. Alternative: single command config
```
  wgconf --interface wg0 \
         --private-key ${wg_private_key} \
         --address 10.0.0.2/24 \
         --peer xTIBA5rboUvnH4htodjb6e6oLhPvYwDhryDYkLpFiGA= \
         --endpoint bootserver.example.com:51820 \
         --allowed-ips 10.0.0.0/24
```
  3. Simple Boot Sequence
```
  #!ipxe
  # Boot script for wireguard-tunneled boot

  # Configure network (get public IP via DHCP)
  dhcp

  # Configure WireGuard tunnel
  set wg0/private-key ${wg_key:uristring}
  set wg0/address 10.0.0.100/24
  set wg0/peer/0/public-key xTIBA5rboUvnH4htodjb6e6oLhPvYwDhryDYkLpFiGA=
  set wg0/peer/0/endpoint boot.netboot.xyz:51820
  set wg0/peer/0/allowed-ips 10.0.0.0/24

  # Activate WireGuard interface
  wgup wg0 || goto retry_wg

  # Now boot through the tunnel
  chain http://10.0.0.1/boot.ipxe || goto retry_boot

  :retry_wg
  echo WireGuard connection failed, retrying...
  sleep 2
  goto wgup

  :retry_boot
  echo Boot failed, retrying...
  sleep 2
  goto chain
```

4. Fun Boot Sequence with "dynamic key distribution" via https
```
  #!ipxe
  # Fetch WireGuard credentials securely, then establish tunnel

  dhcp
  echo Fetching secure boot credentials...

  # Get WireGuard config from provisioning server (over HTTPS)
  imgfetch https://get.massive.wang/wg-config/${net0/mac} --name wgconfig
  imgverify wgconfig https://get.massive.wang/wg-config/${net0/mac}.sig

  # Parse and apply config (would need config parser)
  wgparse wgconfig || goto fail
  wgup wg0 || goto fail

  # Continue boot through tunnel
  chain http://10.0.0.1/boot.ipxe

  :fail
  echo Secure boot failed
  shell
```
